fpgalib/test/test_dac.py

```diff
@@ -0,0 +1,95 @@
+"""This is intended to test fpgalib/dac.py"""
+
+import pytest
+import fpgalib.dac as dac
+
+
+class TestDAC15(object):
+    @classmethod
+    def setup_class(cls):
+        cls.dac = dac.DAC_Build15(0, 'dac_name')
+
+        # sequence SRAM: pi pulse from 0-32, pi pulse from 32-64, ro from 64-400
+        # experiment: first pi -> measure -> second pi -> measure -> end
+        cls.first_pi_start = 0
+        cls.first_pi_end = 32
+        cls.second_pi_start = 32
+        cls.second_pi_end = 64
+        cls.ro_start = 64
+        cls.ro_end = 400
+
+        cls.jump_args = [
+            ('JUMP', [cls.first_pi_end, cls.ro_start, 1]),
+            ('JUMP', [cls.ro_end, cls.second_pi_start, 2]),
+            ('END', [cls.ro_end])
+        ]
+
+    def make_entries(self, jump_args):
+        return [self.dac.make_jump_table_entry(name, args)
+                for name, args in jump_args]
+
+    def test_make_jump_table_entry(self):
+        _ = self.make_entries(self.jump_args)
+        # we made it this far and we're feeling pretty good about it
+
+    def test_make_jump_table(self):
+        entries = self.make_entries(self.jump_args)
+        self.dac.make_jump_table(entries)
+        # we made it this far and we're feeling pretty good about it
+
+    def test_catch_bad_jump_table(self):
+        """Test jump table commands that are executed too frequently"""
+
+        # from_addr of 16 followed by 20 are too close together
+        bad_jump_args = [
+            ('NOP', [16]),
+            ('NOP', [20]),
+            ('END', [400])
+        ]
+
+        # from addr of 16 is followed by from addr of 20
+        bad_jump_args_2 = [
+            ('JUMP', [32, 16, 1]),
+            ('NOP', [20]),
+            ('END', [400])
+        ]
+
+        # from addr of 32 is followed by from addr of 36 (even though IDLE?)
+        bad_jump_args_3 = [
+            ('IDLE', [32, 200]),
+            ('END', [36])
+        ]
+
+        for args in [bad_jump_args, bad_jump_args_2, bad_jump_args_3]:
+            entries = self.make_entries(args)
+            with pytest.raises(AssertionError):
+                self.dac.make_jump_table(entries)
+
+    def test_catch_bad_jump_table_edge(self):
+        """Test a table that executes commands 1 clock cycle too frequently"""
+
+        from_addr = self.dac.JT_MIN_FROM_ADDR * 4
+        bad_jump_args = [
+            ('NOP', [from_addr]),
+            ('NOP', [from_addr +
+                     self.dac.JT_MIN_CLK_CYCLES_BETWEEN_OPCODES * 4]),
+            ('END', [400])
+        ]
+
+        entries = self.make_entries(bad_jump_args)
+        with pytest.raises(AssertionError):
+                self.dac.make_jump_table(entries)
+
+        from_addr = self.dac.JT_MIN_FROM_ADDR * 4
+        bad_jump_args = [
```

Comment:
Again, I would suggest splitting up the good and bad cases into separate, smaller test methods. Also, this table of entries should not be named `bad_*` since it represents a valid jump table.